File: Function_practise/01_random/random-2-2.py
# (2)	產生四個數字，不可以重覆
# 優化
# random.sample 是專門設計用來處理「不重複抽樣」的函式
# random.sample 保證這 P(10,4) 種結果的出現機率嚴格均等

import random

def random_2():
    return "".join(random.sample("0123456789", 4))

print(random_2())

File: Function_practise/01_random/random-2-2.js
//  (2)	產生四個數字，不可以重覆
//  優化

function random2() {
    const digits = "0123456789".split("");

    for (let i = digits.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [digits[i], digits[j]] = [digits[j], digits[i]];
    }

    return digits.slice(0, 4).join("");
}

console.log(random2());

// 解說：流程大概是這樣：
// "0123456789"
//       ↓ split
// ["0", "1", ..., "9"]
//       ↓ Fisher–Yates 洗牌
// ["7", "0", "4", "9", ...]
//       ↓ slice(0, 4)
// ["7", "0", "4", "9"]
//       ↓ join("")
// "7049"